### PR TITLE
Added methods to compute and compare DOS fingerprints

### PR DESCRIPTION
**Pull request by naik-aakash**
##  Compute and compare fingerprints of project and total density of states. 

This functionality could be really useful to compare the DOS of materials from different softwares  (eg:- LOBSTER and VASP ) or versions (Eg:- VASP v5 and VASP v6)

- Added method to compute fingerprints 
- Added method to compare two DOS fingerprints (tanimoto index and dot product)
- Added test cases for the corresponding methods

### CODE REVIEW

File: pymatgen/electronic_structure/dos.py [hunk 1]
```diff
@@ -1165,6 +1166,148 @@ def get_upper_band_edge(
...
+
+        Returns:
+            dict: A dict of the fingerprint Keys=type, Values=np.ndarray(energies, states)
+        """
+        fp_dict = {}
+        fp_dict[fp[2]] = np.array([fp[0], fp[1]]).T
+
+        return fp_dict
+
+    def get_dos_fp_similarity(self, fp1, fp2, col=1, pt="All", normalize=False, tanimoto=False):
```
Comment (JaGeo) [inline pymatgen/electronic_structure/dos.py:1262]: Could this be a static method?

File: pymatgen/electronic_structure/dos.py [hunk 2]
```diff
@@ -1165,6 +1166,148 @@ def get_upper_band_edge(
...
+
+            return fp_tup(np.array([ener]), dos_rebin_sc, type, nbins, bin_width)
+
+        except KeyError:
+            raise ValueError(
+                "Please recheck type requested, either the orbital projections unavailable in input dos or "
+                "some error in the spelling."
+            )
+
+    def _fp_to_dict(self, fp):
```
Comment (JaGeo) [inline pymatgen/electronic_structure/dos.py:1248]: It could be static, I assume

**Review by JaGeo — APPROVED** [review]
You could replace one "finger print" with fingerprint in the documentation but otherwise it looks fine from my side.

**Comment by naik-aakash** [comment]
> You could replace one "finger print" with fingerprint in the documentation but otherwise it looks fine from my side.

Thanks for the review @JaGeo 😄

**Comment by naik-aakash** [comment]
Hi @janosh , This PR is ready to be merge.

File: pymatgen/electronic_structure/dos.py [hunk 3]
```diff
@@ -1165,6 +1166,150 @@ def get_upper_band_edge(
...
+    def get_dos_fp(self, type="summed_pdos", binning=True, min_e=None, max_e=None, nbins=256, normalize=True):
+        """
+        Generates the DOS fingerprint based on work of
+        F. Knoop, T. A. r Purcell, M. Scheffler, C. Carbogno, J. Open Source Softw. 2020, 5, 2671.
+        Source - https://gitlab.com/vibes-developers/vibes/-/tree/master/vibes/materials_fp
+        Copyright (c) 2020 Florian Knoop, Thomas A.R.Purcell, Matthias Scheffler, Christian Carbogno
+
+
+        Args:
+            type (str): Specify fingerprint type needed can accept 's/p/d/f/summed_pdos/tdos'
```
Comment (janosh) [inline pymatgen/electronic_structure/dos.py:1178]: Please specify the default values for each keyword argument in the doc string.

File: pymatgen/electronic_structure/dos.py [hunk 4]
```diff
@@ -1165,6 +1166,150 @@ def get_upper_band_edge(
...
+            min_e (float): The minimum mode energy to include in the fingerprint
+            max_e (float): The maximum mode energy to include in the fingerprint
+            nbins (int): Number of bins to be used in the fingerprint
```
Comment (janosh) [inline pymatgen/electronic_structure/dos.py:1181]: Can we rename this to `n_bins` for proper snake casing?

**Review by janosh — CHANGES_REQUESTED** [review]
Thanks @naik-aakash! I left some comments.

File: pymatgen/electronic_structure/dos.py (hunk 3 above)
Comment (naik-aakash) [reply pymatgen/electronic_structure/dos.py:1186]: Thanks for the review @janosh , I made the requested changes.

File: pymatgen/electronic_structure/dos.py (hunk 4 above)
Comment (naik-aakash) [reply pymatgen/electronic_structure/dos.py]: Done 😃

File: pymatgen/electronic_structure/dos.py [hunk 5]
```diff
@@ -1165,6 +1166,150 @@ def get_upper_band_edge(
         upper_band_edge = energies[np.argmax(densities)]
         return upper_band_edge
 
+    def get_dos_fp(self, type="summed_pdos", binning=True, min_e=None, max_e=None, nbins=256, normalize=True):
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply pymatgen/electronic_structure/dos.py]: Thanks for the suggestions , I have made the requested changes.

**Comment by naik-aakash** [comment]
Hi @janosh , all suggestions have been incorporated now. And tests seems to pass as well. 😄

File: pymatgen/electronic_structure/dos.py [hunk 6]
```diff
@@ -1165,6 +1166,166 @@ def get_upper_band_edge(
...
+        if max_e is None:
+            max_e = np.max(energies)
+
+        if min_e is None:
+            min_e = np.min(energies)
+
+        pdos_obj = self.get_spd_dos()
+
+        pdos = {}
+        for k, _v in pdos_obj.items():
```
Comment (janosh) [inline pymatgen/electronic_structure/dos.py:1209]:
```suggestion
        for key in pdos_obj:
```

File: pymatgen/electronic_structure/dos.py [hunk 7]
```diff
@@ -1165,6 +1166,166 @@ def get_upper_band_edge(
...
+            return fingerprint(np.array([ener]), dos_rebin_sc, type, n_bins, bin_width)
+
+        except KeyError:
+            raise ValueError(
+                "Please recheck type requested, either the orbital projections unavailable in input dos or "
+                "some there exist some mistake in the spelling."
+            )
+
+    @staticmethod
+    def fp_to_dict(fp):
```
Comment (janosh) [inline pymatgen/electronic_structure/dos.py:1258]: Missing type annotation

File: pymatgen/electronic_structure/dos.py [hunk 8]
```diff
@@ -1165,6 +1166,166 @@ def get_upper_band_edge(
...
 
+    def get_dos_fp(
+        self,
+        type: str = "summed_pdos",
+        binning: bool = True,
+        min_e: float | None = None,
+        max_e: float | None = None,
+        n_bins: int = 256,
+        normalize: bool = True,
+    ):
```
Comment (janosh) [inline pymatgen/electronic_structure/dos.py:1177]: Missing return type.

File: pymatgen/electronic_structure/dos.py (hunk 7 above)
Comment (naik-aakash) [reply pymatgen/electronic_structure/dos.py:1258]: I have added the type annotation (namedtuple) here now.

File: pymatgen/electronic_structure/dos.py (hunk 8 above)
Comment (naik-aakash) [reply pymatgen/electronic_structure/dos.py:1177]: I have added the type annotation (namedtuple) here now.

**Comment by naik-aakash** [comment]
Hi @janosh , thank you for the really useful tips and suggestions. 😃

**Comment by naik-aakash** [comment]
Hi @janosh , I do not know why so many tests are failing. I looked at Failed tests and they do not have anything to do with this PR. I hope it could still be merged.

**Comment by janosh** [comment]
@naik-aakash Sorry for the delay here. Back from traveling now. PR looks good! Made a few last changes. Failures like you said are unrelated. Ready to merge.

**Review by janosh — APPROVED** [review]

**Comment by naik-aakash** [comment]
> @naik-aakash Sorry for the delay here. Back from traveling now. PR looks good! Made a few last changes. Failures like you said are unrelated. Ready to merge.

@janosh No worries! Thank you. Happy Holidays

**Pull request merged**